tiledb/api/src/datatype/arrow.rs

```diff
@@ -46,169 +53,667 @@ impl ArrowPrimitiveTypeNative for f64 {
+pub enum DatatypeToArrowResult {
+    /// There is an exact logical match for the tiledb `Datatype`.
+    /// The individual values of the respective types have the same bit width
+    /// and are meant to be interpreted the same way.
+    ///
+    /// In general, this means that:
+    /// 1. `CellValNum::Fixed(1)` maps to an arrow primitive or date/time type.
+    /// ```
+    /// use tiledb::{array::CellValNum, datatype::arrow::DatatypeToArrowResult};
+    /// assert_eq!(DatatypeToArrowResult::Exact(arrow::datatypes::DataType::UInt8),
+    ///            tiledb::datatype::arrow::to_arrow(&tiledb::Datatype::UInt8, CellValNum::single()));
+    /// ```
+    /// 2. `CellValNum::Fixed(n) if n > 1` 1 maps to an arrow fixed size list.
+    /// ```
+    /// use arrow::datatypes::DataType as Arrow;
+    /// use tiledb::{Datatype as TileDB, array::CellValNum, datatype::arrow::DatatypeToArrowResult};
+    /// let arrow = tiledb::datatype::arrow::to_arrow(&TileDB::UInt8, CellValNum::try_from(8).unwrap());
+    /// let DatatypeToArrowResult::Exact(Arrow::FixedSizeList(item, fixed_len)) = arrow else { unreachable!() };
+    /// assert_eq!(*item.data_type(), Arrow::UInt8);
+    /// assert_eq!(fixed_len, 8);
+    /// ```
+    /// 3. `CellValNum::Var` maps to an arrow `LargeList`.
+    /// ```
+    /// use arrow::datatypes::DataType as Arrow;
+    /// use tiledb::{Datatype as TileDB, array::CellValNum, datatype::arrow::DatatypeToArrowResult};
+    /// let arrow = tiledb::datatype::arrow::to_arrow(&TileDB::UInt8, CellValNum::Var);
+    /// let DatatypeToArrowResult::Exact(Arrow::LargeList(item)) = arrow else { unreachable!() };
+    /// assert_eq!(*item.data_type(), Arrow::UInt8);
+    /// ```
+    ///
+    /// There are some exceptions, such as `(Datatype::Blob, CellValNum::Var)`
+    /// mapping to `arrow::datatypes::DataType::LargeBinary`, which is always variable-length.
+    ///
+    /// ```
+    /// use tiledb::{array::CellValNum, datatype::arrow::DatatypeToArrowResult};
+    /// assert_eq!(DatatypeToArrowResult::Exact(arrow::datatypes::DataType::LargeBinary),
+    ///            tiledb::datatype::arrow::to_arrow(&tiledb::Datatype::Blob, CellValNum::Var));
+    /// ```
+    /// When the output is any kind of list, field metadata may be used to represent the exact
+    /// input datatype if the input on its own is an inexact match.
+    /// ```
+    /// use arrow::datatypes::DataType as Arrow;
+    /// use tiledb::{Datatype as TileDB, array::CellValNum, datatype::arrow::DatatypeToArrowResult};
+    /// use tiledb::datatype::arrow::{to_arrow, ARROW_FIELD_METADATA_KEY_TILEDB_EXACT_MATCH};
+    /// let arrow = to_arrow(&TileDB::StringAscii, CellValNum::Var);
+    /// let DatatypeToArrowResult::Exact(Arrow::LargeList(item)) = arrow else { unreachable!() };
+    /// assert_eq!(*item.data_type(), Arrow::UInt8);
+    /// let Some(s) = item.metadata().get(ARROW_FIELD_METADATA_KEY_TILEDB_EXACT_MATCH)
+    /// else { unreachable!() };
```

Comment:
Huh. TIL: https://doc.rust-lang.org/rust-by-example/flow_control/let_else.html

Reply:
Also the lack of formatting of doc strings makes me sad.

Reply:
Sounds about Rust: It does exist, its just disabled by default and unstable. https://rust-lang.github.io/rustfmt/?version=v1.5.1&search=format_code_in_doc_comments#format_code_in_doc_comments